packages/website/blog/2024-03-25-changes-to-consistent-type-imports-with-decorators.md

```diff
@@ -0,0 +1,101 @@
+  - and the import **_IS_** annotated as `import type` then TS will emit metadata that annotates the type of `arg` as `Function`.
+
+## `consistent-type-imports` caused runtime breakage
+
+The important piece is that last dot point - the handling of imported names that resolve to class declarations. If the import is not annotated as `import type` then TS emits a runtime reference to the imported name. This runtime reference is implicit and requires type information to derive - you cannot derive its existence purely based on single-file AST analysis.
+
+The `consistent-type-imports` rule was introduced to allow users to enforce that any imported names are annotated as `import type` if they are not used in a value location. How the rule makes this decision is based on single-file AST analysis; it scans the code using a technique called scope analysis so that it can find all references to the imported names and determine if each reference is a value reference.
+
+The issue arises with legacy decorators and decorator metadata - syntactically the only reference to `Foo` is a type reference. However the emitted code contains a hidden reference to `Foo`. When the rule relies upon the code it sees then it will report an error and attempt to mark `Foo` as `import type`. If the user applies this fix then that will cause their runtime code to change (`arg`'s metadata goes from `Foo` to `Function`) which can have downstream runtime impacts and cause broken code!
+
+## Past (Broken) Solution
+
+In the past we tried to solve this problem by enforcing that imported names that are used in decorator metadata are specifically _not_ marked as type-only imports to ensure that values are always correctly emitted in the runtime code.
+
+However this solution had a hidden pitfall; if the user also used `isolatedModules: true` then TS will enforce that all imported types are explicitly marked as `import type` for compatibility with single-file build tools. This lead to an unresolvable situation where `consistent-type-imports` would enforce that an imported name _must not be_ marked with `import type` so that we could ensure we don't break decorator metadata, and simultaneously TS would enforce that that same imported name _must be_ marked with `import type`.
+
+There have been a few attempts to fix this issue but the resolution we came to was that the only solution was to add type information to the rule so that it could correctly understand all of the above type-aware constraints. Adding type information to an existing rule is something we try to avoid because it is a major breaking change that restricts the rule to just users that leverage [type-aware linting](/getting-started/typed-linting).
+
+However we decided that adding type-information to the rule to handle this edge case was not a positive change for users or the ecosystem for two main reasons:
```

Comment:
Being a little more precise here: we didn't decide reality, we just determined it.

```suggestion
Adding type-information to the rule to handle this edge case would not be a positive change for users or the ecosystem:
```